### PR TITLE
Expose getEditedPostAttribute selector

### PR DESCRIPTION
**Pull request by abotteram**
## Description

In https://github.com/WordPress/gutenberg/pull/4802 @aduth voiced some concerns if attributes should have dedicated selectors for the `select` API. This PR removes the `getEditedPostSlug` and `getEditedPostTitle` selectors and exposes the `getEditedPostAttribute` selector.

I have removed the `getEditedPostContent` selector from the public API but not the selector itself, because it contains additional logic. I added a switch statement to `getEditedPostAttribute` to call `getEditedPostContent` when the attribute name passed is `content`.

Initially also added `getEditedPostAttributes` selector this PR, but moved it to a separate branch. `add/get-edited-post-attributes-selector`.

## How Has This Been Tested?

Rewritten existing unit tests.
Console tested.

## Types of changes

Breaking change, if someone has already used the `getEditedPostSlug` or `getEditedPostTitle` selector.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Review by atimmer — APPROVED** [review]
Maybe, `getEditedPostAttribute` should call `getEditedPostContent` if it is called with 'content' as the second argument. What do you think @youknowriad and @gziolo?

**Comment by gziolo** [comment]
> Maybe, getEditedPostAttribute should call getEditedPostContent if it is called with 'content' as the second argument. What do you think @youknowriad and @gziolo?

We can also remove `getEditedPostContent` as it seems to be obsolete with the new unified approach.

**Comment by aduth** [comment]
Unlike others, `getEditedPostContent` is not simply a property accessor proxy and has its own logic: Specifically that it serializes blocks in most cases.

Whether it should be exposed depends on whether we care to expose this detail to the consumer (i.e. that it has custom logic), or if from a developer's perspective they only care that they receive content and not so much how it is generated.

**Comment by gziolo** [comment]
@aduth, I was educated by @Xyfi about this complex logic of `getEditedPostContent`, but I still think that it makes sense to have one selector that knows how to return all types of attributes. As a plugin developer, I shouldn't be forced to learn that those attributes are handled differently. In the future, we can also make their life easier if we decide to support returning a group of attributes like this:

```
const applyQuery = query( ( select ) => ( {
	...select( 'core/editor', 'getEditedPostAttribute', [ 'title', 'content', 'slug' ] ),
} ) );
```

**Comment by aduth** [comment]
That sounds fine to me.

**Comment by Xyfi** [comment]
I could quite easily implement that in this PR. But I would rather see a separate function to get multiple attributes, which would lead to a cleaner codebase.

**Comment by gziolo** [comment]
> I could quite easily implement that in this PR. But I would rather see a separate function to get multiple attributes, which would lead to a cleaner codebase.

Yes, it makes sense 👍

File: edit-post/components/sidebar/featured-image/index.js [hunk 1]
```diff
@@ -43,7 +43,7 @@ function FeaturedImage( { isOpened, postType, onTogglePanel } ) {
 }
 
 const applyQuery = query( ( select ) => ( {
-	postTypeSlug: select( 'core/editor', 'getCurrentPostType' ),
+	postTypeSlug: select( 'core/editor', 'getEditedPostAttribute' ),
```
Comment (gziolo) [inline edit-post/components/sidebar/featured-image/index.js]: `type` should be passed as the last argument.

File: editor/store/test/selectors.js [hunk 2]
```diff
@@ -26,9 +26,7 @@ const {
 	getCurrentPostLastRevisionId,
 	getCurrentPostRevisionsCount,
 	getCurrentPostType,
```
Comment (gziolo) [inline editor/store/test/selectors.js]: Should we also remove this one in the follow-up PR?

Comment (gziolo) [reply editor/store/test/selectors.js]: `select( 'core/editor', 'getEditedPostAttribute', 'type' )` - we use that already in this PR.

**Comment by aduth** [comment]
This might be related to the failing tests 😅 

image

**Comment by aduth** [comment]
Steps to reproduce:

1. Try switching to Text mode

File: editor/store/selectors.js [hunk 3]
```diff
@@ -239,11 +228,39 @@ export function getPostEdits( state ) {
...
+ * saved state of the post.
+ *
+ * @param {Object}   state          Global application state.
+ * @param {string[]} attributeNames Post attribute name.
+ *
+ * @return {Object}  Post attributes mapped to object by name.
+ */
+export function getEditedPostAttributes( state, attributeNames ) {
+	const attributes = {};
+	map( attributeNames, attributeName => {
```
Comment (aduth) [inline editor/store/selectors.js]: This is pretty prime candidate for `reduce`:

```js
export function getEditedPostAttributes( state, attributeNames ) {
	return reduce( attributeNames, ( result, attributeName ) => {
		const value = getEditedPostAttribute( state, attributeName );
		if ( value ) {
			result[ attributeName ] = value;
		}

		return result;
	}, {} );
}
```

**Review by gziolo — APPROVED** [review]
It seems like code where I left comment: https://github.com/WordPress/gutenberg/pull/5003/files#r167861481, shouldn't be in this PR. Otherwise it is ready to go. 👍

**Comment by gziolo** [comment]
Thanks for performing this refactoring. I like how flexible this new selector is. Do we have those selectors documented? It would be nice to start promoting them in Gutenberg handbook.

**Comment by Shelob9** [comment]
@gziolo I created #5033 to improve the docs for those selectors - this change, which is really cool, made one of the examples there incorrect. I documented allow the post attributes I could find -post, type, slug and content. Are there others I can use? 

Am I right to assume that `wp.data.select( 'core/editor', 'getEditedPostAttribute', 'excerpt' )` would provide the post excerpt? I can't figure out -- but also, I'm not fully caffeinated -- if any post attributes are available via that selector or only a specific white list. I'd like to improve that example.

**Comment by gziolo** [comment]
> Am I right to assume that `wp.data.select( 'core/editor', 'getEditedPostAttribute', 'excerpt' )` would provide the post excerpt? I can't figure out -- but also, I'm not fully caffeinated -- if any post attributes are available via that selector or only a specific white list. I'd like to improve that example.

We would have to test it, you can check inside the state. I would document only those attributes that we use in Gutenberg at the moment, maybe even add a whitelist to have a better overview what we need to support in the future.

**Comment by aduth** [comment]
One issue I have here is that it's my understanding with the split between `editor` and `edit-post` (#4661) that we'll want all post-specific behaviors (including selectors) to live in `edit-post`.

Currently this is exposed under the `editor` module key, and I expect this will break in a future release.

I might suggest that we either move forward with migrating more of these selectors to `edit-post`, or artificially expose them in the correct namespace (i.e. keep in `editor/store/index.js`, but call `registerSelectors` with `'core/edit-post'` as the namespace for selectors we know to be post-specific).

cc @youknowriad

**Comment by youknowriad** [comment]
>  might suggest that we either move forward with migrating more of these selectors to edit-post, or artificially expose them in the correct namespace (i.e. keep in editor/store/index.js, but call registerSelectors with 'core/edit-post' as the namespace for selectors we know to be post-specific).

I don't have a precise idea how this will evolve. I think the refactoring we plan to do for the editor to consider it as a template editor defaulting to `post title - post content` and making the editor generic enough to handle editing with or without post might have an impact on the need or the shape of these selectors. So my personal opinion for now would be to just stick with what we have right now but granted that this will create some backwards compatibility burden.

**Pull request merged**